openedx_learning/core/components/models.py

```diff
@@ -0,0 +1,295 @@
+
+    A Component belongs to one and only one LearningPackage.
+
+    The UUID should be treated as immutable. The identifier field *is* mutable,
+    but changing it will affect all ComponentVersions.
+    """
+
+    uuid = immutable_uuid_field()
+    learning_package = models.ForeignKey(LearningPackage, on_delete=models.CASCADE)
+
+    # namespace and type work together to help figure out what Component needs
+    # to handle this data. A namespace is *required*. The namespace for XBlocks
+    # is "xblock.v1" (to match the setup.py entrypoint naming scheme).
+    namespace = models.CharField(max_length=100, null=False, blank=False)
+
+    # type is a way to help sub-divide namespace if that's convenient. This
+    # field cannot be null, but it can be blank if it's not necessary. For an
+    # XBlock, type corresponds to tag, e.g. "video". It's also the block_type in
+    # the UsageKey.
+    type = models.CharField(max_length=100, null=False, blank=True)
+
+    # identifier is local to a learning_package + namespace + type. For XBlocks,
+    # this is the block_id part of the UsageKey, which usually shows up in the
+    # OLX as the url_name attribute.
+    identifier = identifier_field()
+
+    created = manual_date_time_field()
+    modified = manual_date_time_field()
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(
+                fields=[
+                    "learning_package",
+                    "namespace",
+                    "type",
+                    "identifier",
+                ],
+                name="component_uniq_lc_ns_type_identifier",
+            )
+        ]
+        verbose_name = "Component"
+        verbose_name_plural = "Components"
+
+    def __str__(self):
+        return f"{self.identifier}"
+
+
+class ComponentVersion(models.Model):
+    """
+    A particular version of a Component.
+
+    A new ComponentVersion should be created anytime there is either a change to
+    the content or a change to the policy around a piece of content (e.g.
+    schedule change).
+
+    Each ComponentVersion belongs to one and only one Component.
+    """
+
+    uuid = immutable_uuid_field()
+    component = models.ForeignKey(Component, on_delete=models.CASCADE)
+    title = models.CharField(max_length=1000, default="", null=False, blank=True)
+    version_num = models.PositiveBigIntegerField(
+        null=False,
+        validators=[MinValueValidator(1)],  # Versions start with 1
+    )
+    created = manual_date_time_field()
+
+    # For later consideration:
+    # created_by = models.ForeignKey(
+    #    settings.AUTH_USER_MODEL,
+    #    # Don't delete content when the user who created it had their account removed.
+    #    on_delete=models.SET_NULL,
+    #
+
+    contents = models.ManyToManyField(
+        "Content",
+        through="ComponentVersionContent",
+        related_name="component_versions",
+    )
```

Comment:
Huh... I guess it's possible that we'd make one without Content (like just a title). I'd rather make the admin read-only altogether though, because ComponentVersion creation should include adding Publish entries rather than just direct model manipulation (I guess I really should do the api.py pieces of this soon).